### PR TITLE
Modify PyTorch's integration of NNPACK to use a unified underlying thread pool implementation.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27402 Use NNPACK for strided convolutions.
* **#27341 Modify PyTorch's integration of NNPACK to use a unified underlying thread pool implementation.**

Multi-threaded:

```
Pixel 2:
Before:    362.716
PR-27402:  185.799
PR-27341:  142.011

Pixel 3:
Before:    246.755
PR-27402:  160.045
PR-27341:  115.437

```

Single-threaded:

```
Pixel 2:
Before:    308.084
PR-27340:  303.539
PR-27341:  313.558

Pixel 3:
Before:    234.272
PR-27340:  227.158
PR-27341:  232.787

```

Differential Revision: [D17835333](https://our.internmc.facebook.com/intern/diff/D17835333)

### CODE REVIEW

File: aten/src/ATen/native/NNPACK.cpp [hunk 1]
```diff
@@ -56,53 +56,39 @@ bool _nnpack_available() {
...
-    }
-  }
-  return nnpack_threadpool_;
+  });
+
+  return nnpack_successfully_initialized;
+}
+
+static pthreadpool_t nnpack_threadpool() {
+  return caffe2::mobile_pthreadpool();
```
Comment (ljk53) [inline aten/src/ATen/native/NNPACK.cpp]: mobile_pthreadpool() returns null for non-mobile build.
I don't know if NNPACK is used for non-mobile build at all - maybe @ezyang / @ilia-cher knows?
if we want to keep 100% the same behavior we should fallback to old thread creation logic for non-mobile build...

File: aten/src/ATen/native/NNPACK.cpp [hunk 2]
```diff
@@ -142,7 +128,7 @@ constexpr int weight_width_dim = 3;
 // Often written as 2 + max_dim (extra dims for batch size and channels)
 // constexpr int max_dim = 3;
 
-std::vector<int64_t> conv_output_size(
+static std::vector<int64_t> conv_output_size(
```
Comment (ljk53) [inline aten/src/ATen/native/NNPACK.cpp]: nit: I'd do this in a separate PR (as we might want to request pick to 1.3 we should keep this as small as possible).

**Review by ljk53 — COMMENTED** [review]
Thanks for discovering and fixing this problem!

File: aten/src/ATen/native/NNPACK.cpp (hunk 2 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/NNPACK.cpp]: Good point, definitely.

File: aten/src/ATen/native/NNPACK.cpp (hunk 1 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/NNPACK.cpp]: Right, valid point.  I'll wait for Illia's input.

Comment (ezyang) [reply aten/src/ATen/native/NNPACK.cpp]: This test suggest nnpack is used in non-mobile:

```
    @unittest.skipIf(not torch._nnpack_available(), "NNPACK unavailable")
    def test_nnpack_conv(self):
        for kern, inp_size in [(3, 6), (3, 7), (4, 9)]:
            for batch, padding, chan_in, chan_out in \
                    product([1, 2], [0, 1, 2], [2], [3]):
```

Comment (ilia-cher) [reply aten/src/ATen/native/NNPACK.cpp]: let's fallback to a non-mobile thread pool

File: aten/src/ATen/native/NNPACK.cpp [hunk 3]
```diff
@@ -55,53 +55,67 @@ bool _nnpack_available() {
...
+    const uint32_t threads = at::get_num_threads();
 #else
-    threads = std::thread::hardware_concurrency();
+    const uint32_t threads = std::thread::hardware_concurrency();
 #endif
+
     nnpack_threadpool_ = pthreadpool_create(threads);
-    if (nnpack_threadpool_ == nullptr) {
-      throw std::runtime_error("could not initialize NNPack's pthreadpool");
+    if ( !nnpack_threadpool_ ) {
```
Comment (ljk53) [inline aten/src/ATen/native/NNPACK.cpp]: nit: remove extra space inside '()'

**Review by ljk53 — APPROVED** [review]

File: aten/src/ATen/native/NNPACK.cpp (hunk 3 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/NNPACK.cpp]: Fixed it in 27402 to avoid having this PR rebuilt.
